### PR TITLE
dev-util/edi: new package

### PR DESCRIPTION
**Pull request by juippis**
Tested with clang-6 and clang-7. I'm open to suggestions for handling the clang check better, tried with toolchain-funcs but couldn't get anything smart out of that.

Closes: https://bugs.gentoo.org/604932

### CODE REVIEW

File: dev-util/edi/edi-0.7.0.ebuild [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+	clang? ( sys-devel/clang:* )
+"
+DEPEND="
+	${RDEPEND}
+	dev-libs/check
+"
+
+src_prepare() {
+	default
+	xdg_environment_reset
```
[the remark this answers is not in the document]
Comment (juippis) [reply dev-util/edi/edi-0.7.0.ebuild]: Actually, looks like meson.eclass has been updated to EAPI-7 a week ago.

File: dev-util/edi/edi-0.7.0.ebuild [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+}
+
+src_configure() {
+	local emesonargs=(
+		-D bear=false
+	)
+
+	if use clang; then
+		local clangver=$(clang --version | grep -Eo '[0-9]' | head -n1) || die
```
[the remark this answers is not in the document]
Comment (juippis) [reply dev-util/edi/edi-0.7.0.ebuild]: So, something like:
`llvm-config --version | grep -Eo '[0-9]' | head -n1`
Then use `get_llvm_prefix` to get the libdir and include dir? I'll need to look into llvm.eclass when I get home. Will also check the license and `usex` thing then.

[the remark this answers is not in the document]
Comment (mgorny) [reply dev-util/edi/edi-0.7.0.ebuild]: No, use the eclass to find llvm-config, then get --includedir and --libdir from it.

File: dev-util/edi/edi-0.7.0.ebuild [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+
+EAPI=6
+
+inherit meson xdg-utils
+
+DESCRIPTION="An EFL-based IDE"
+HOMEPAGE="https://www.enlightenment.org/about-edi"
+SRC_URI="https://github.com/Enlightenment/edi/archive/v${PV}.tar.gz -> ${P}.tar.gz"
+
+LICENSE="GPL-2 LGPL-2.1"
```
[the remark this answers is not in the document]
Comment (juippis) [reply dev-util/edi/edi-0.7.0.ebuild]: https://github.com/Enlightenment/edi/blob/develop/COPYING.GPLv2
https://github.com/Enlightenment/edi/blob/develop/COPYING.LGPL

Once again, to my best knowledge, these are not +. **COPYING.GPLv2** is identical to */usr/portage/licenses/GPL-2*, however **COPYING.LGPL** seems to have a linking exception, 
https://github.com/Enlightenment/edi/blob/develop/COPYING.LGPL#L233
so I guess this is **LGPL-2.1-with-linking-exception**

File: dev-util/edi/edi-0.7.0.ebuild (hunk 2 above)
[the remark this answers is not in the document]
Comment (juippis) [reply dev-util/edi/edi-0.7.0.ebuild]: Without testing, is this the correct idea: 
```
local xyz=$(get_llvm_prefix "${LLVM_MAX_SLOT}")/bin/llvm-config

-D libclang-headerdir=$(xyz --includedir)
-D libclang-libdir=$(xyz --libdir)
```

**Comment by juippis** [comment]
Pushing a new version in a moment. 
 - EAPI is still 6 because *llvm.eclass* needs it, 
 - I used `clang? ( sys-devel/clang:${LLVM_SLOT}= )` not sure if it's correct or not, 
 - fixed QA issues with `sed`, 
 - I am unsure with the syntax used in `if use clang;... fi` part

It builds and works with clang and -clang.

File: dev-util/edi/edi-0.7.0.ebuild (hunk 3 above)
[the remark this answers is not in the document]
Comment (mgorny) [reply dev-util/edi/edi-0.7.0.ebuild]: Use the source(s), Luke.

File: dev-util/edi/edi-0.7.0.ebuild (hunk 2 above)
[the remark this answers is not in the document]
Comment (mgorny) [reply dev-util/edi/edi-0.7.0.ebuild]: `pkg_setup` in the eclass will pull correct llvm-config into the PATH.

File: dev-util/edi/edi-0.7.0.ebuild (hunk 3 above)
[the remark this answers is not in the document]
Comment (mgorny) [reply dev-util/edi/edi-0.7.0.ebuild]: Hmm, it looks that the app is supposed to have command-line options for 'copyright' and 'license'. Could you see what they do? It really sucks that they don't even say how the two licenses interact in README.

File: dev-util/edi/edi-0.7.0.ebuild [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+SRC_URI="https://github.com/Enlightenment/edi/archive/v${PV}.tar.gz -> ${P}.tar.gz"
+
+LICENSE="GPL-2 LGPL-2.1-with-linking-exception"
+SLOT="0"
+KEYWORDS="~amd64 ~x86"
+IUSE="clang"
+
+RDEPEND="
+	>=dev-libs/efl-1.21.1[eet,X]
+	clang? ( sys-devel/clang:${LLVM_SLOT}= )
```
Comment (mgorny) [inline dev-util/edi/edi-0.7.0.ebuild]: Where does this `LLVM_SLOT` thing come from? oO

File: dev-util/edi/edi-0.7.0.ebuild [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+"
+DEPEND="
+	${RDEPEND}
+	dev-libs/check
+"
+
```
Comment (mgorny) [inline dev-util/edi/edi-0.7.0.ebuild]: You're missing `llvm_check_deps()` to ensure that your selected slot actually includes clang.

File: dev-util/edi/edi-0.7.0.ebuild [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+src_configure() {
+	local emesonargs=(
+		-D bear=false
+		-D libclang=$(usex clang true false)
+	)
+
+	if use clang; then
+		local llvmconfbin=$(get_llvm_prefix "${LLVM_SLOT}")/bin/llvm-config
+		emesonargs+=(
+			-D libclang-headerdir="$($llvmconfbin --includedir || die)"
```
Comment (mgorny) [inline dev-util/edi/edi-0.7.0.ebuild]: Variables are always `${foo}` form. Also, you can call it via PATH.

File: dev-util/edi/metadata.xml [hunk 7]
```diff
@@ -0,0 +1,14 @@
...
+<pkgmetadata>
+  <maintainer type="person">
+    <email>juippis@gmail.com</email>
+  </maintainer>
+  <maintainer type="project">
+    <email>proxy-maint@gentoo.org</email>
+    <name>Proxy Maintainers</name>
+  </maintainer>
+  <use>
+    <flag name="clang">Provide integration with <pkg>sys-devel/clang</pkg> for autocompletion and inline errors</flag>
```
Comment (mgorny) [inline dev-util/edi/metadata.xml]: Line wrapping?

**Comment by juippis** [comment]
Addressed all previous comments. Limited *metadata.xml* to 90 chars, because thats where packages.gentoo.org seems to wrap it too. 

It's become rather complicated to what it was at first, but hopefully getting there ;)

**Comment by mgorny** [comment]
```
>>> Test phase: dev-util/edi-0.7.0
ninja -v -j12 -l0 -C /tmp/portage/dev-util/edi-0.7.0/work/edi-0.7.0-build test
ninja: Entering directory `/tmp/portage/dev-util/edi-0.7.0/work/edi-0.7.0-build'
[0/1] /usr/lib/python-exec/python3.7/meson -u test --no-rebuild --print-errorlogs
usage: meson [-h] [--buildtype {plain,debug,debugoptimized,release,minsize,custom}] [--strip] [--unity {on,off,subprojects}] [--prefix PREFIX]
             [--libdir LIBDIR] [--libexecdir LIBEXECDIR] [--bindir BINDIR] [--sbindir SBINDIR] [--includedir INCLUDEDIR] [--datadir DATADIR]
             [--mandir MANDIR] [--infodir INFODIR] [--localedir LOCALEDIR] [--sysconfdir SYSCONFDIR] [--localstatedir LOCALSTATEDIR]
             [--sharedstatedir SHAREDSTATEDIR] [--werror] [--warnlevel {1,2,3}] [--layout {mirror,flat}] [--default-library {shared,static,both}]
             [--backend {ninja,vs,vs2010,vs2015,vs2017,xcode}] [--stdsplit] [--errorlogs] [--install-umask INSTALL_UMASK]
             [--auto-features {enabled,disabled,auto}] [--optimization {0,g,1,2,3,s}] [--debug] [-D option] [--cross-file CROSS_FILE] [-v]
             [--wrap-mode {default,nofallback,nodownload,forcefallback}] [--fatal-meson-warnings] [--reconfigure]
             [builddir] [sourcedir]
meson: error: unrecognized arguments: -u --no-rebuild --print-errorlogs
FAILED: meson-test 
/usr/lib/python-exec/python3.7/meson -u test --no-rebuild --print-errorlogs
ninja: build stopped: subcommand failed.

**Comment by juippis** [comment]
Ok, try now :)

File: dev-util/edi/edi-0.7.0.ebuild [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+	>=dev-libs/efl-1.21.1[eet,X]
+	clang? ( sys-devel/clang:= )
+"
+DEPEND="
+	${RDEPEND}
+	dev-libs/check
+"
+BDEPEND="virtual/pkgconfig"
+
+RESTRICT="test"
```
Comment (mgorny) [inline dev-util/edi/edi-0.7.0.ebuild]: RESTRICT=test requires an explanatory comment saying what's wrong with tests.

**Comment by floppym** [comment]
The testing error was caused by a regression in meson-0.48.0. Please upgrade to >=dev-util/meson-0.48.0-r2.

**Comment by juippis** [comment]
Yeah, upgrading meson and removing test restriction does finish the merge. Gonna remove it and rebase, since I see that "faulty" meson versions were removed from the tree.

**Comment by mgorny** [comment]
I'll merge, thanks!

**Pull request closed without merging**